### ISSUE TITLE
vtable corruption in Python specialization of C++ classes

### ISSUE THREAD

**Issue by pikacic** [root-project/root#6376]
## Describe the bug
Using cppyy in ROOT master (51eb56e), I have a Python class that inherits from a C++ class (with two abstract classes) and I need to pass an instance of the Python class to a C++ function that accepts a pointer to one of the base classes.

When the C++ function tries to call a method from the instance, I get a segmentation fault (actually I enter in an infinite segfault loop, where ROOT segfault handler segfaults too).

## Expected behavior
Just that called the C++ function can use the passed instance.

## To Reproduce
From a directory containing the file:
```reproducer.py
import cppyy

cppyy.gbl.gInterpreter.Declare("""
#include <array>
#include <iostream>

struct Interface1 {
  virtual void do_1()   = 0;
  virtual ~Interface1() = default;
};

struct Interface2 {
  virtual void do_2()   = 0;
  virtual ~Interface2() = default;
};

struct Base : virtual public Interface1, virtual public Interface2 {
  Base() {
    std::cout << std::hex << "Base* " << this << "\nInt1* " << static_cast<Interface1*>( this ) << "\nInt2* "
              << static_cast<Interface2*>( this ) << '\n';
  }
};

struct Derived : Base, virtual public Interface2 {
  void do_1() override { std::cout << std::hex << this << "->do_1\n"; }
  void do_2() override { std::cout << std::hex << this << "->do_2\n"; }
};

void my_func( Interface2* i ) { i->do_2(); }
""")


class PyDerived(cppyy.gbl.Derived):
    pass


i = PyDerived()
i.do_1()
i.do_2()

cppyy.gbl.my_func(i)

```
call
```
python reproducer.py
```

## Setup
- ROOT version: 51eb56e
- OS: CentOS7
- arch: x86_64
- compiler: gcc 9.2.0

## Additional context
I'm using the nightly builds by CERN SFT.

**Comment by pikacic** [comment]
You can find the output of valgrind attached: [valgrind.txt](https://github.com/root-project/root/files/5240343/valgrind.txt)

**Comment by etejedor** [comment]
Thank you @pikacic for the reproducer and the valgrind report, much appreciated! I made a PR to fix the issue, it was just merged. I will check the nightlies tomorrow just in case.

**Comment by etejedor** [comment]
Fix backported to 6.22 here: https://github.com/root-project/root/pull/6450
